drivers/SmartThings/matter-switch/src/switch_utils/utils.lua

```diff
@@ -166,16 +168,64 @@ function utils.component_to_endpoint(device, component)
   return utils.find_default_endpoint(device)
 end
 
-function utils.endpoint_to_component(device, ep)
-  local map = device:get_field(fields.COMPONENT_TO_ENDPOINT_MAP) or {}
-  for component, endpoint in pairs(map) do
-    if endpoint == ep then
+--- @param device any a Matter device object
+--- @param opts number|table either is an ep_id or a table { endpoint_id, capability_id }
+--- @return string component
+function utils.endpoint_to_component(device, opts)
+  local ep_info = {}
+  if type(opts) == "number" then
+    ep_info.endpoint_id = opts
+  elseif type(opts) == "table" then
+    if opts.endpoint_info then
+      ep_info = opts.endpoint_info
+    else
+      ep_info = {
+        endpoint_id = opts.endpoint_id,
+        cluster_id = opts.cluster_id,
+        attribute_id = opts.attribute_id
+      }
+    end
+  end
+  for component, map_info in pairs(device:get_field(fields.COMPONENT_TO_ENDPOINT_MAP) or {}) do
+    if type(map_info) == "number" and map_info == ep_info.endpoint_id then
       return component
+    elseif type(map_info) == "table" and map_info.endpoint_id == ep_info.endpoint_id then
+      if (not map_info.cluster_id or (map_info.cluster_id == ep_info.cluster_id
+        and utils.tbl_contains(map_info.attribute_ids, ep_info.attribute_id)))
+        and (not opts.capability_id or utils.tbl_contains(map_info.capability_ids, opts.capability_id)) then
+        return component
+      end
     end
   end
   return "main"
 end
 
+--- @param device any a Matter device object
+--- @param ep_info number|table endpoint_id or ib (includes endpoint_id, cluster_id, attribute_id)
+--- @param event any a capability event object
+function utils.emit_event_for_endpoint(device, ep_info, event)
```

Comment:
nit: could you add a few sentences above this function explaining why it is needed and what overrides it provides on top of the existing library `emit_event_for_endpoint` function? Could you do the same for the `endpoint_to_component` override as well?